Add glossary to manual (#1143)
seL4/seL4: manual/parts/glossary.tex

%
% Copyright 2014, General Dynamics C4 Systems
%
% SPDX-License-Identifier: GPL-2.0-only
%

\newglossaryentry{asid}{
    name=ASID,

>     description={Address Space Identifier. ASIDs are associated with page directories (PDs), and define the virtual address space of a thread. Multiple threads may be in the same address space}

Reviewer: ```suggestion
    description={Address Space Identifier. Depending on architecture, the kernel provides software ASIDs, which are associated with VSpace root objects, and define the virtual address space of a thread. They are mapped to hardware ASIDs on demand when the architecture supports these. Multiple threads may be in the same address space}
```

}

\newglossaryentry{badge}{
    name=Badge,

>     description={A badge is a piece of extra information stored in a capability. It can be used by applications to identify caps previously handed out to clients}

Reviewer: ```suggestion
    description={A badge is a piece of extra information stored in a capability, mostly used in endpoint and notification capabilities. It can be used by applications to identify caps previously handed out to clients}
```

Reviewer 2: "with" or "for" seems more appropriate than "in". Otherwise a good improvement. Any other use cases for badges are truly obscure and I can't think of anything useful, other than perhaps limiting further copying or revocation.

Reviewer: Yup, happy with "for"

Author: I made the change using:
description={A badge is a piece of extra information stored in a capability, mostly used for endpoint and notification capabilities. It can be used by applications to identify caps previously handed out to clients}

}

\newglossaryentry{cap}{
    name=Capability,
    description={The main access control concept in seL4. A capability conceptually is a reference to a kernel object together with a set of access rights. Most seL4 capabilities store additional bits of information. Some of this additional information may be exposed to the user, but the bulk of it is kernel-internal book-keeping information. Capabilities are stored in CNodes and TCBs}
}

\newglossaryentry{cdt}{
    name=CDT,

>     description={Capability Derivation Tree. A kernel-internal data structure that tracks the child/parent relationship between capabilities. Capabilities to new objects are children of the Untyped capability the object was created from. Capabilities can also be copied; in this case the user may specify if the operation should produce children or siblings of the source capability. The revoke operation will delete all children of the invoked capability}

Reviewer: ```suggestion
    description={Capability Derivation Tree. A kernel-internal data structure that tracks the child/parent relationship between capabilities. Capabilities to new objects are children of the Untyped capability the object was created from. Capabilities can also be copied and result in either child or sibling capabilities, depending on the operation that was used and the depth of the existing derivation tree. The revoke operation will delete all children of the invoked capability}
```

}

\newglossaryentry{cnode}{
    name=CNode,
    description={Capability Node. Kernel-controlled storage that holds capabilities. Capability nodes can be created in different sizes and be shared between CSpaces}
}

\newglossaryentry{cptr}{
    name=CPtr,
    description={Capability Pointer. A user-level reference to a capability, relative to a specified root CNode or the thread’s CSpace root}
}

\newglossaryentry{cspace}{
    name=CSpace,

>     description={A directed graph of CNodes. The CSpace of a thread defines the set of capabilities it has access to. The root of the graph is the CNode capability in the CSpace slot of the thread. The edges of the graph are the CNode capabilities residing in the CNodes .spanned by this root}

Reviewer: ```suggestion
    description={A directed graph of CNodes. The CSpace of a thread defines the set of capabilities it has access to. The root of the graph is the CNode capability in the CSpace slot of the thread. The edges of the graph are the CNode capabilities residing in the CNodes spanned by this root}
```

}

\newglossaryentry{ipc}{
    name=IPC,
    description={Inter Process Communication is facilitated by endpoints, which act as general communication ports. Invocations on endpoint objects are used to send and receive messages}
}

\newglossaryentry{irqcontrol}{
    name=IRQControl,
    description={A single capability from which IRQHandler capabilities to all IRQ numbers in the system can be derived. This capability can be moved between CSpaces and CSlots but cannot be duplicated. Revoking this capability removes all IRQHandlers}
}

\newglossaryentry{irqhandler}{
    name=IRQHandler,
    description={Capabilities that represent the ability of a thread to handle a certain interrupt. See \autoref{ch:io}}
}

\newglossaryentry{notificationobject}{
    name=Notification Object,
    description={A word-size array of flags that provides a non-blocking signalling mechanism similar to a binary semaphore. Operations are signalling a subset of flags in a single operation, polling to check any flags, and blocking until any are signalled. Notification capabilities can be signal-only or wait-only}
}

\newglossaryentry{replyobject}{

>     name=Reply Object,
>     description={A reply object is a vessel for tracking reply messages, used to send a reply message and
>     wake up the caller}

Reviewer: ```suggestion
    name=Reply Object,
    description={(MCS only) A reply object is a vessel for tracking reply messages, used to send a reply message and
    wake up the caller}
```

}

\newglossaryentry{schedulingcontext}{
    name=Scheduling Context,
    description={(MCS only) An abstraction of CPU execution time}
}

\newglossaryentry{tcb}{
    name=TCB,
    description={Thread Control Block. The kernel object that stores management data for threads, such as the thread's CSpace, VSpace, thread state, or user registers}
}

\newglossaryentry{untypedmemory}{

>     name=Untyped Memory,
>     description={The foundation of memory allocation in the seL4 kernel. See \autoref{sec:kernmemalloc}}

Reviewer: ```suggestion
    name=Untyped Memory,
    description={Memory that can be used to create kernel objects or user-accessible frames. The foundation of memory allocation in the seL4 kernel. See \autoref{sec:kernmemalloc}}
```

Reviewer 2: Strictly speaking it can only be used to create kernel objects, except if Page objects are not kernel objects, in which case that should be mentioned explicitly (instead of frame). I would go for:

"The foundation of kernel object creation and memory allocation in the seL4 kernel."

Reviewer: It's correct that frames are also kernel objects, the difference I was trying to point out is that these are directly writeable by user space, and the rest are not. Happy to tweak that. What I'm not happy with in the "foundation" sentence before was that it doesn't say what you can do with these objects. That untyped memory is the foundation of something doesn't help the reader to understand where to go with that. Maybe we should just point at the retype operation. (I'm happy mentioning the foundation bit, it should just not be the only thing).

Reviewer: You're right, the whole user-writeable thing is leading in the wrong direction. Happy to remove that.

Author: @lsf37 @Indanz I'm getting a bit bogged down in details :-)
What wording would you like to use for the Description of Untyped Memory?

Reviewer: I'd go for

```suggestion
    name=Untyped Memory,
    description={Memory that can be used to create kernel objects via the \apifunc{seL4\_Untyped\_Retype}{untyped_retype} invocation. It is the foundation of memory allocation in the seL4 kernel. See \autoref{sec:kernmemalloc}}
```
